File: src/main/java/com/dec/decisland/client/model/ClothesModel.java
package com.dec.decisland.client.model;// Made with Blockbench 5.0.7
// Exported for Minecraft version 1.17 or later with Mojang mappings
// Paste this class into your mod and generate all required imports


import com.dec.decisland.DecIsland;
import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.blaze3d.vertex.VertexConsumer;
import net.minecraft.client.model.HumanoidModel;
import net.minecraft.client.model.geom.ModelLayerLocation;
import net.minecraft.client.model.geom.ModelPart;
import net.minecraft.client.model.geom.PartPose;
import net.minecraft.client.model.geom.builders.*;
import net.minecraft.client.renderer.entity.state.HumanoidRenderState;
import net.minecraft.resources.Identifier;

public class ClothesModel<T extends HumanoidRenderState> extends HumanoidModel<T> {
	// This layer location should be baked with EntityRendererProvider.Context in the entity renderer and passed into this model's constructor
	public static final ModelLayerLocation LAYER_LOCATION = new ModelLayerLocation(Identifier.fromNamespaceAndPath(DecIsland.MOD_ID, "clothes_model"), "main");

	public ClothesModel(ModelPart root) {
        super(root);
	}

	@Override
	public void setupAnim(T state) {
		// 重置所有部分的变换
		this.head.xRot = 0;
		this.head.yRot = 0;
		this.head.zRot = 0;
		this.body.xRot = 0;
		this.body.yRot = 0;
		this.body.zRot = 0;
		this.rightArm.xRot = 0;
		this.rightArm.yRot = 0;
		this.rightArm.zRot = 0;
		this.leftArm.xRot = 0;
		this.leftArm.yRot = 0;
		this.leftArm.zRot = 0;
		this.rightLeg.xRot = 0;
		this.rightLeg.yRot = 0;
		this.rightLeg.zRot = 0;
		this.leftLeg.xRot = 0;
		this.leftLeg.yRot = 0;
		this.leftLeg.zRot = 0;
	}

	public static LayerDefinition createBodyLayer() {
		MeshDefinition meshdefinition = new MeshDefinition();
		PartDefinition partdefinition = meshdefinition.getRoot();

		// 添加head部分
		PartDefinition head = partdefinition.addOrReplaceChild("head", CubeListBuilder.create()
			.texOffs(0, 0).addBox(-4.0F, -8.0F, -4.0F, 8.0F, 8.0F, 8.0F, new CubeDeformation(0.0F)),
			PartPose.offset(0.0F, 0.0F, 0.0F));

		// 添加hat部分作为head的子部分
		head.addOrReplaceChild("hat", CubeListBuilder.create()
			.texOffs(32, 0).addBox(-4.0F, -8.0F, -4.0F, 8.0F, 8.0F, 8.0F, new CubeDeformation(0.5F)),
			PartPose.offset(0.0F, 0.0F, 0.0F));

		// 添加body部分
		PartDefinition body = partdefinition.addOrReplaceChild("body", CubeListBuilder.create()
			.texOffs(16, 20).addBox(-4.0F, 0.0F, -2.0F, 8.0F, 12.0F, 4.0F, new CubeDeformation(0.0F)),
			PartPose.offset(0.0F, 0.0F, 0.0F));

		// 添加手臂部分
		PartDefinition right_arm = partdefinition.addOrReplaceChild("right_arm", CubeListBuilder.create()
			.texOffs(40, 37).addBox(-3.0F, -2.0F, -2.0F, 4.0F, 12.0F, 4.0F, new CubeDeformation(0.0F)),
			PartPose.offset(-5.0F, 2.0F, 0.0F));
		PartDefinition left_arm = partdefinition.addOrReplaceChild("left_arm", CubeListBuilder.create()
			.texOffs(32, 53).addBox(-1.0F, -2.0F, -2.0F, 4.0F, 12.0F, 4.0F, new CubeDeformation(0.0F)),
			PartPose.offset(5.0F, 2.0F, 0.0F));

		// 添加腿部部分
		PartDefinition right_leg = partdefinition.addOrReplaceChild("right_leg", CubeListBuilder.create()
			.texOffs(0, 37).addBox(-2.0F, 0.0F, -2.0F, 4.0F, 12.0F, 4.0F, new CubeDeformation(0.0F)),
			PartPose.offset(-1.9F, 12.0F, 0.0F));
		PartDefinition left_leg = partdefinition.addOrReplaceChild("left_leg", CubeListBuilder.create()
			.texOffs(16, 53).addBox(-2.0F, 0.0F, -2.0F, 4.0F, 12.0F, 4.0F, new CubeDeformation(0.0F)),
			PartPose.offset(1.9F, 12.0F, 0.0F));

		return LayerDefinition.create(meshdefinition, 64, 64);
	}
}

File: src/main/java/com/dec/decisland/client/model/ClothesModel.kt
package com.dec.decisland.client.model

import com.dec.decisland.DecIsland
import com.dec.decisland.client.bedrock.model.BedrockBone
import com.dec.decisland.client.bedrock.model.BedrockCube
import com.dec.decisland.client.bedrock.model.BedrockEntityAssets
import net.minecraft.client.model.HumanoidModel
import net.minecraft.client.model.geom.ModelLayerLocation
import net.minecraft.client.model.geom.ModelPart
import net.minecraft.client.model.geom.PartPose
import net.minecraft.client.model.geom.builders.CubeListBuilder
import net.minecraft.client.model.geom.builders.LayerDefinition
import net.minecraft.client.model.geom.builders.MeshDefinition
import net.minecraft.client.renderer.entity.state.HumanoidRenderState
import net.minecraft.resources.Identifier
import kotlin.math.PI

class ClothesModel<T : HumanoidRenderState>(root: ModelPart) : HumanoidModel<T>(root) {
    override fun setupAnim(state: T) {
        super.setupAnim(state)
    }

    companion object {
        @JvmField
        val LAYER_LOCATION: ModelLayerLocation =
            ModelLayerLocation(Identifier.fromNamespaceAndPath(DecIsland.MOD_ID, "clothes_model"), "main")

        private val GEOMETRY_LOCATION: Identifier =
            Identifier.fromNamespaceAndPath(DecIsland.MOD_ID, "bedrock/models/armor/clothes.geometry.json")

        @JvmStatic
        fun createBodyLayer(): LayerDefinition {
            val geometry = BedrockEntityAssets.geometry(GEOMETRY_LOCATION)
            val meshDefinition = MeshDefinition()
            val root = meshDefinition.root

            val head = root.addOrReplaceChild(
                "head",
                cubeListFromBone(geometry.bonesByName.getValue("armorHead")),
                poseFromBone(geometry.bonesByName.getValue("armorHead")),
            )
            head.addOrReplaceChild(
                "hat",
                CubeListBuilder.create(),
                PartPose.offset(0.0f, 0.0f, 0.0f),
            )

            root.addOrReplaceChild(
                "body",
                cubeListFromBone(geometry.bonesByName.getValue("armorBody")),
                poseFromBone(geometry.bonesByName.getValue("armorBody")),
            )
            root.addOrReplaceChild(
                "right_arm",
                cubeListFromBone(geometry.bonesByName.getValue("armorRightArm")),
                poseFromBone(geometry.bonesByName.getValue("armorRightArm")),
            )
            root.addOrReplaceChild(
                "left_arm",
                cubeListFromBone(geometry.bonesByName.getValue("armorLeftArm")),
                poseFromBone(geometry.bonesByName.getValue("armorLeftArm")),
            )
            root.addOrReplaceChild(
                "right_leg",
                cubeListFromBone(geometry.bonesByName.getValue("armorRightLeg")),
                poseFromBone(geometry.bonesByName.getValue("armorRightLeg")),
            )
            root.addOrReplaceChild(
                "left_leg",
                cubeListFromBone(geometry.bonesByName.getValue("armorLeftLeg")),
                poseFromBone(geometry.bonesByName.getValue("armorLeftLeg")),
            )

            return LayerDefinition.create(meshDefinition, geometry.textureWidth, geometry.textureHeight)
        }

        private fun cubeListFromBone(bone: BedrockBone): CubeListBuilder {
            val builder = CubeListBuilder.create()
            bone.cubes.forEach { cube ->
                appendCube(builder, bone, cube)
            }
            return builder
        }

        private fun appendCube(builder: CubeListBuilder, bone: BedrockBone, cube: BedrockCube) {
            val uv = cube.uv
            if (uv != null) {
                builder.texOffs(uv.first, uv.second)
            }
            builder.mirror(cube.mirror)
            builder.addBox(
                cube.origin.x - bone.pivot.x,
                -(cube.origin.y - bone.pivot.y) - cube.size.y,
                cube.origin.z - bone.pivot.z,
                cube.size.x,
                cube.size.y,
                cube.size.z,
                net.minecraft.client.model.geom.builders.CubeDeformation(cube.inflate),
            )
        }

        private fun poseFromBone(bone: BedrockBone): PartPose = PartPose.offsetAndRotation(
            bone.pivot.x,
            24.0f - bone.pivot.y,
            bone.pivot.z,
            bone.rotation.x.toModelRadX(),
            bone.rotation.y.toModelRadY(),
            bone.rotation.z.toModelRadZ(),
        )

        private fun Float.toRadians(): Float = (this * PI.toFloat()) / 180.0f

        private fun Float.toModelRadX(): Float = -toRadians()

        private fun Float.toModelRadY(): Float = -toRadians()

        private fun Float.toModelRadZ(): Float = toRadians()
    }
}
